apps/client/src/app/pages/inventory-optimizer/optimizations/unwanted-materials.ts

```diff
@@ -0,0 +1,88 @@
+import { InventoryOptimizer } from './inventory-optimizer';
+import { InventoryItem } from '../../../model/user/inventory/inventory-item';
+import { UserInventory } from '../../../model/user/inventory/user-inventory';
+import { ListRow, getItemSource } from '../../../modules/list/model/list-row';
+import { LazyDataService } from '../../../core/data/lazy-data.service';
+import { environment } from '../../../../environments/environment';
+
+export class UnwantedMaterials extends InventoryOptimizer {
+
+  static RECIPE_ILVL_KEY = 'optimizer:unwanted-materials:recipe-ilvl';
+  static UNWANTED_CACHE_KEY = 'optimizer:unwanted-materials:cache';
+
+  constructor(private lazyData: LazyDataService) {
+    super();
+
+    const appVersion = environment.version;
+    const cache = JSON.parse(localStorage.getItem(UnwantedMaterials.UNWANTED_CACHE_KEY));
+    if (!cache || appVersion !== cache.version) {
+      this.rebuildCache(appVersion);
+      //console.log("Rebuilt Cache");
+    } 
+    else {
+      //console.log("Cache version matches");
+    }
+  }
+
+  //This just caches a list of materials used in all recipes, including ilvl, and company workshop usage
+  rebuildCache(version: string) {
+    const recipes = this.lazyData.data.recipes;
+    const cache = JSON.parse(localStorage.getItem(UnwantedMaterials.UNWANTED_CACHE_KEY)) || {};
+    cache["version"] = version;
+    cache.materials = cache.materials || {};
+
+    //console.log(recipes[0]);
+
+    recipes.forEach(recipe => {
+      const firstIlvl = this.lazyData.data.ilvls[recipe.result];
+
+      //We make a recursive function so we can get all the way to base materials to update their ilvls
+      const spanAllMaterials = (ingredientId: number, ilvl: number) => {        
+        //Grab the material from the cache if it already exists
+
+        const materialIlvl = cache.materials[ingredientId];
+        if (materialIlvl) {
+          //if this new recipe had a higher ilvl, we update the cache
+          const newilvl = this.lazyData.data.ilvls[ingredientId];
```

Comment:
As `this.lazyData.data.ilvls` is of type `any`, the `newilvl` variable is of type `any` too, but we know it's number, so better set it so Typescript can do more type checking.
```suggestion
          const newilvl: number = this.lazyData.data.ilvls[ingredientId];
```